# Variant index notes

Main.scala
- Commands.variantIndex()
- Calls VariantIndex.builder(configuration).build
- Saves result in parquet format.

VariantIndex.scala
- call Builder.build
    - `raw` is basic dataframe based on `conf.variantIndex.raw`
    - `nearests`:
        - Uses `Distance.scala` to calculate tss distance
        - Dependent on `GeneIndex.BioTypes`. This should be refactored and possible placed in config.
        - Creates `GeneIndex` which reads in json file `conf.ensembl.lut`
            - the `lut` file is generated by a python script `create_genes_dictionary`.
            - the only fields we're missing is `tss` which we can calculate using the existing fields in the target 
              index.
                - `df['tss'] = df.apply(lambda row: row['start'] if row['fwdstrand'] else row['end'], axis=1)`
                - in Spark `df.withColumn("tss", when(col("fwdstrand") === 1, col("start").otherwise(col("end"))`

## Final output

- Entries with linked gene 72858944 as indicated by `SELECT count(*) FROM `open-targets-genetics-dev.genetics_dev.variants` as v where v.gene_id_any is not null;`
- Entries with no gene_id: 19765

# Variant-gene notes

`Main.variantToGene()`
    - loads variantIndex
    - computes distance
        - this has a dependency on `conf.ensembl.lut`. The exact same calculation is done in the `variant-index`.
    - computes 4 `V2GIndex.Components`
    - calls V2GIndex.build
    - saves in `configuration.format` format under `configuration.variantGene.path`

